### PR TITLE
[DI] Allow auto-configured method calls.

### PR DESCRIPTION
**Pull request by GaryPEGEOT**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    |no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Allow to auto-configure method calls like:
```php
$container->registerForAutoconfiguration(LoggerAwareInterface::class)->addMethodCall('setLogger', array(new Reference(LoggerInterface::class)));
```

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php [hunk 1]
```diff
@@ -63,7 +60,10 @@ private function processDefinition(ContainerBuilder $container, $id, Definition
 
         $definition->setInstanceofConditionals(array());
         $parent = $shared = null;
-        $instanceofTags = array();
+        $instanceof = array(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php]: another variable instead of an array would be better I think

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php [hunk 2]
```diff
@@ -77,11 +77,13 @@ private function processDefinition(ContainerBuilder $container, $id, Definition
             foreach ($instanceofDefs as $key => $instanceofDef) {
                 /** @var ChildDefinition $instanceofDef */
                 $instanceofDef = clone $instanceofDef;
-                $instanceofDef->setAbstract(true)->setParent($parent ?: 'abstract.instanceof.'.$id);
-                $parent = 'instanceof.'.$interface.'.'.$key.'.'.$id;
+                $instanceofDef->setAbstract(true)->setParent($parent ?: "abstract.instanceof.$id");
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php]: please revert these changes, they are unrelated (did fabbot suggest them? if yes, please it :) )

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php [hunk 3]
```diff
@@ -91,23 +93,24 @@ private function processDefinition(ContainerBuilder $container, $id, Definition
 
         if ($parent) {
             $bindings = $definition->getBindings();
-            $abstract = $container->setDefinition('abstract.instanceof.'.$id, $definition);
+            $abstract = $container->setDefinition("abstract.instanceof.$id", $definition);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php]: same as above

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveInstanceofConditionalsPassTest.php [hunk 4]
```diff
@@ -60,7 +60,6 @@ public function testProcessInheritance()
         (new ResolveChildDefinitionsPass())->process($container);
 
         $expected = array(
-            array('foo', array('bar')),
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveInstanceofConditionalsPassTest.php]: changing an existing test case is always suspicious,
is this really legitimate? won't it introduce BC breaks?
we need to answer these questions carefully

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/integration/defaults_instanceof_importance/expected.yml [hunk 5]
```diff
@@ -20,7 +20,6 @@ services:
         # calls from instanceof are kept, but this comes later
         calls:
             # first call is from instanceof
-            - [setSunshine, [bright]]
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/integration/defaults_instanceof_importance/expected.yml]: same as above: changing an existing test case needs to be doubly confirmed

**Review by nicolas-grekas — COMMENTED** [review]
thanks for starting this work

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php (hunk 2 above)
Comment (GaryPEGEOT) [reply src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php]: Not fabbot, just did it because string interpolation run faster in PHP7, but I can revert it.

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveInstanceofConditionalsPassTest.php (hunk 4 above)
Comment (GaryPEGEOT) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveInstanceofConditionalsPassTest.php]: This change came from the fact that if a definition has already a method call (line 125: `if (!$definition->hasMethodCall($call[0])) {`), I do not add it again to avoid call a setter multiple times. Hard to say if we should deduplicate or not method calls...

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php (hunk 1 above)
Comment (GaryPEGEOT) [reply src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php]: Fixed in `b61d981`

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/ResolveInstanceofConditionalsPass.php]: yes please do, that's unrelated

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveInstanceofConditionalsPassTest.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveInstanceofConditionalsPassTest.php]: > Hard to say if we should deduplicate or not method calls
we should not:
```
$twig->addExtension($ext1);
$twig->addExtension($ext1);
...
```

Comment (GaryPEGEOT) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveInstanceofConditionalsPassTest.php]: I removed method filtering. Method order has changed, but seems normal since we now allow calls inheritance and they are added at the end of the process.

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/integration/defaults_instanceof_importance/expected.yml [hunk 6]
```diff
@@ -19,8 +19,7 @@ services:
             - { name: from_defaults }
         # calls from instanceof are kept, but this comes later
         calls:
-            # first call is from instanceof
-            - [setSunshine, [bright]]
-            #
             - [enableSummer, [true]]
             - [setSunshine, [warm]]
+            # Call from instanceof
```
Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/integration/defaults_instanceof_importance/expected.yml]: if the order of calls changes, that means a BC break

**Comment by GaryPEGEOT** [comment]
Methods calls are now in the right order, error on Travis is unrelated (error with RedisAdapter).

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveInstanceofConditionalsPassTest.php (hunk 4 above)
Comment (GaryPEGEOT) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveInstanceofConditionalsPassTest.php]: Test order fixed in `da6d7e4`

**Comment by xabbuh** [comment]
@GaryPEGEOT Why did you close here?

**Comment by nicolas-grekas** [comment]
Continued in #26768

**Comment by GaryPEGEOT** [comment]
Lost some of my commits, so openned a new one
